### DISCUSSION TITLE
Identity provider token exchange

### DISCUSSION THREAD

**Discussion by todor-susnjevic** [keycloak/keycloak#22806] [Q&A] [answered]
I am building API that is secured by keycloak. Users can be registered with email and pass and after that can get access token. Now I want to enable users to login with google, facebook, apple. I'm was able to do token exchange. Now I want to cover case where I already have a user on keycloak registered with email/password but now wants to login with google. If try token exchange I will get error: federated_identity_account_exists. How to implements this?

**Comment by justin-tay** [comment]
For that particular requirement you should have configured Keycloak as an identity broker as documented at https://www.keycloak.org/docs/latest/server_admin/#_identity_broker instead of using token exchange. That way when the users use different identity providers to sign in with the same e-mail address they will be given the opportunity to do account linking.

![identity-providers](https://github.com/keycloak/keycloak/assets/49700559/6451bde2-0d4f-4673-8e38-833853ba27a5)

**Reply by todor-susnjevic** [reply]
I do not understand you answer. Yes, i have configured google as idp. The idea is, as I understand, mobile user to authenticate with google (or facebook, apple..) and get google access token. For that google access token I can get with token exchange keyclaok access token that mobile client can use to access the API. And that case is working. My question is what to do if I already have some non social linked account on keycloak, or account from other IdP like twitter? I configured first broker login flow to send email about linking but it's not coming. Can I maybe programatically link accounts with same email? Thanks.
<img width="1175" alt="Screenshot 2023-08-30 at 16 45 57" src="https://github.com/keycloak/keycloak/assets/31806412/5651fd5e-acef-4916-bfb4-44d2ce9aca44">


And smtp email is configured:


<img width="1920" alt="Screenshot 2023-08-30 at 16 52 12" src="https://github.com/keycloak/keycloak/assets/31806412/124b7c7b-659c-49c3-b933-91ab89a523ea">

What should trigger that account linking email to be send? I am trying to do token exchange as in the case then account with same email not exist and it gives me error
2023-08-30 16:52:03,158 WARN [org.keycloak.events] (executor-thread-44) type=TOKEN_EXCHANGE_ERROR, realmId=5fdc0765-5e60-4460-8483-244e3039917b, clientId=rori-client, userId=null, ipAddress=127.0.0.1, error=federated_identity_account_exists, auth_method=token_exchange, grant_type=urn:ietf:params:oauth:grant-type:token-exchange, subject_issuer=google, validation_method='user info', client_auth_method=client-secret.

**Reply by justin-tay** [reply]
That's not really the correct flow. I don't think the first broker login flow is used when you are doing token exchange. Some of the code in the `DefaultTokenExchangeProvider` looks similar to some of the authenticators used in that flow though.

The authorization endpoint is queried using a browser user agent, so in that case a user can interact with the identity provider. The token endpoint, which you use to do your token exchange, is queried using the client application, for which the user cannot interact.

The flow is for your mobile application, using the AppAuth library, to navigate to Keycloak's authorization endpoint first. The user then can choose which mode to login as. If he chooses Facebook, Keycloak will then broker to Facebook. If that e-mail is already used then Keycloak can ask the user to do account linking.

![identity-provider-login-page](https://github.com/keycloak/keycloak/assets/49700559/d619c664-b7e0-4d41-9772-af5fff3ef6b9)

**Reply by todor-susnjevic** [reply]
That is not a solution, I can't change look for the mobile apps. I need to control that account linking on my backend programaticaly.

**Reply by todor-susnjevic** [reply]
And if I am not suppose to use token exchange, how can mobile application or web site client can get access token from keycloak to assess my backend API, in the case that user is logged with some social network?

**Reply by justin-tay** [reply]
Yeah so what you want really isn't the way it is supposed to work but if you insist you can write your own implementation of the token exchange provider.

The following is the logic that is doing the import.

https://github.com/keycloak/keycloak/blob/3c00286e6cd51cba843b5e08aff07ddc83e28a74/services/src/main/java/org/keycloak/protocol/oidc/DefaultTokenExchangeProvider.java#L519-L624

**Reply by justin-tay** [reply, marked as answer]
> And if I am not suppose to use token exchange, how can mobile application or web site client can get access token from keycloak to assess my backend API, in the case that user is logged with some social network?

The normal flow is the standard authorization code flow to query Keycloak's authorization endpoint, it will then broker to the social network if needed. Your application then uses the code to get the tokens at Keycloak's token endpoint.

You can read [RFC 8252 - OAuth 2.0 for Native Apps](https://datatracker.ietf.org/doc/html/rfc8252) for how it's supposed to work https://datatracker.ietf.org/doc/html/rfc8252#section-4.1.
